src/model_vit.py

```diff
@@ -231,12 +236,67 @@ def predict_step(
 
-        # Save embeddings in GeoParquet format
+        return gdf
+
+    def on_predict_epoch_end(self) -> gpd.GeoDataFrame:
+        """
+        Logic to gather all the results from one epoch in a prediction loop.
+
+        This is where we combine all the vector embeddings generated from each
+        mini-batch prediction (stored in geopandas.GeoDataFrame tables) in a
+        row-wise manner. The embeddings are then output to GeoParquet files
+        according to their MGRS code.
+
+        Steps:
+        1. Concatenate all geopandas.GeoDataFrame objects row-wise
+        2. Find all unique MGRS names in the big GeoDataFrame.
+        3. Split the GeoDataFrame by MGRS name, and output each MGRS-subsetted
+           table to a GeoParquet file.
+            ________________     ________________     ________________
+           |                |   |                |   |                |
+           | GeoDataFrame 1 |   | GeoDataFrame 2 |   | GeoDataFrame 3 |
+           | for MGRS 12ABC |   | for MGRS 23DEF |   | for MGRS 45GHI |
+           |________________|   |________________|   |________________|
+                    |                    |                    |
+                    v                    v                    v
+               12ABC_v01.gpq        23DEF_v01.gpq        45GHI_v01.gpq
+        """
+        # Combine list of geopandas.GeoDataFrame objects
+        results: list[gpd.GeoDataFrame] = self.trainer.predict_loop.predictions
+        if results:
+            gdf: gpd.GeoDataFrame = pd.concat(
+                objs=results, axis="index", ignore_index=True
+            )
+        else:
+            print(
+                "No embeddings generated, "
+                f"possibly no GeoTIFF files in {self.trainer.datamodule.data_path}"
+            )
+            return
+
+        # Save embeddings in GeoParquet format, one file for each MGRS code
         outfolder: str = f"{self.trainer.default_root_dir}/data/embeddings"
         os.makedirs(name=outfolder, exist_ok=True)
-        outpath = f"{outfolder}/embeddings_{batch_idx}.gpq"
-        gdf.to_parquet(path=outpath, schema_version="1.0.0")
-        print(f"Saved embeddings of shape {tuple(embeddings_mean.shape)} to {outpath}")
+
+        # Find unique MGRS names (e.g. '12ABC'), e.g.
+        # from 's3://.../.../claytile_12ABC_20201231_v02_0001.tif', get 12ABC
+        mgrs_codes = gdf.source_url.str.split("/").str[-1].str.split("_").str[1]
+        unique_mgrs_codes = mgrs_codes.unique()
+        for mgrs_code in unique_mgrs_codes:
+            if re.match(pattern=r"(\d{2}[A-Z]{3})", string=mgrs_code) is None:
+                raise ValueError(
+                    "MGRS code should have 2 numbers and 3 letters (e.g. 12ABC), "
+                    f"but got {mgrs_code} instead"
+                )
+
+            # Output to a GeoParquet filename like {MGRS:5}_v{VERSION:2}.gpq
+            outpath = f"{outfolder}/{mgrs_code}_v01.gpq"
```

Comment:
@mattpaul, are you ok with a filename like `12ABC_v01.gpq`? We could also add a prefix like `embedding_` or `MGRS_` if you prefer.

Reply:
This convention looks good. Will all the chips for a given MGRS tile be contained in a single GeoParquet file?

Reply:
Cool, I'll stick with that filename convention then!

> Will all the chips for a given MGRS tile be contained in a single GeoParquet file? Assuming something like date or build number goes into `outfolder` ?

Yes, each file (e.g. `32VLM_v01.gpq`) would contain all the embeddings for every chip within that MGRS tile. It will also contain multiple dates, so you could have overlapping chips in any one spatial area, due to images taken at different dates. Easiest way might be to show you how this looks like in QGIS:

![Peek 2023-12-13 18-19](https://github.com/Clay-foundation/model/assets/23487320/0c2b76eb-9ecb-4166-a5fe-89cbd31a1eab)

I've put 75% transparency for each green chip, and you'll notice that some chip areas are lighter in colour (so only 1 embedding from 1 date), whereas others are darker in colour from multiple overalps (2 embeddings from 2 dates).

Not sure what you mean by build number. Do you mean the `_v01` part?

Reply:
OK, cool, I see... the animated gif helps, thanks!

@weiji14 from our previous discussion around having `{DATE}`  in the file name I thought that embeddings from the same tile/chip geometry taken at different dates would be serialized into separate vector embedding files which might be preferable considering we may want to _incrementally_ add to the set of vector embeddings over time.

@leothomas thoughts re: pros vs cons of combining multiple embeddings from different dates / times into a single GeoParquet export file per tile across all time?  I don't know if we plan to ever want to train the model on all 7 years of available Sentinel-2 data but as a thought experiment for the sake of argument that would yield a rather large file that increases in size over time. Contrast that with an incremental approach in which the number of GeoParquet file objects in S3 would grow over time as the model learns more and more data but the individual file size is expected to remain static which makes it easier to scale ingestion horizontally.

re: build number - I am thinking about the model training lifecycle. If we zoom out so to speak and consider Clay's roadmap over time w.r.t. model training considering the availability of new Sentinel data products, etc. we will likely want to re-train the model periodically over time. 

Perhaps once a month or once a quarter we may want to refresh the training dataset with the latest data available from Sentinel-1,2 or perhaps train on new sensor data.

We will want to be able to distinguish between old builds or old model version vs new versions because parts of our distributed system will need to continue working with old version of data, embeddings and map image tile assets while the next version is being processed.

Hence we'll want something like a _build number_ or _model version_ to uniquely identify each along the way. [Similar to how OpenAI has tagged versions of GPT](https://platform.openai.com/docs/models/continuous-model-upgrades), etc.

I assumed that `_v01` was meant to track our embedding file format (which has already changed quite a few times if I'm not mistaken), or was that intended to track model version?

Reply:
> @weiji14 from our previous discussion around having `{DATE}` in the file name I thought that embeddings from the same tile/chip geometry taken at different dates would be serialized into separate vector embedding files which might be preferable considering we may want to _incrementally_ add to the set of vector embeddings over time.
> 
> @leothomas thoughts re: pros vs cons of combining multiple embeddings from different dates / times into a single GeoParquet export file per tile across all time? I don't know if we plan to ever want to train the model on all 7 years of available Sentinel-2 data but as a thought experiment for the sake of argument that would yield a rather large file that increases in size over time. Contrast that with an incremental approach in which the number of GeoParquet file objects in S3 would grow over time as the model learns more and more data but the individual file size is expected to remain static which makes it easier to scale ingestion horizontally.

Just to do some quick math. One embedding file for an MGRS tile (roughly 3 dates, ~1000 rows in total) is about 3.0MB. Let's say there are 73 dates in a year (365/5-day revisit), and 7 years (2017-2023), so $3.0\text{MB} \times 73 \times 7 = 1533\text{MB} = 1.5\text{GB}$. This is likely an overestimate since Parquet's columnar nature allows for good compression, but let's assume 1.5GB, is that too big a file to ingest? We'll have 56984 MGRS tiles for the whole world, so that would be $1.5\text{GB} \times 56984\text{files} = 85.5\text{TB}$ of embeddings in total to process (again, likely a gross overestimate, as we won't process everything).

What about if we named the file like `{MGRS}_{MINDATE}_{MAXDATE}_{VERSION}.gpq`, e.g. `32VLM_20170517_20221119_v01.gpq`? Storing 1 date per file would result in too many files, but we can store a range of dates at least (e.g. annual, every two years, etc). Using a range-based `{MINDATE}_{MAXDATE}` naming scheme would:

1. Allow someone to quickly find the relevant files to process based on a YYYYMMDD date just by looking at the filenames, and
2. Be fairly compatible with any changes in the range of dates we want to store in one file, compared to say, a naming scheme with just one date like `{YYYYMMDD}`.


> re: build number - I am thinking about the model training lifecycle. If we zoom out so to speak and consider Clay's roadmap over time w.r.t. model training considering the availability of new Sentinel data products, etc. we will likely want to re-train the model periodically over time.
> 
> Perhaps once a month or once a quarter we may want to refresh the training dataset with the latest data available from Sentinel-1,2 or perhaps train on new sensor data as well in addition to the Sentinel data products.
> 
> We will want to be able to distinguish between old builds or old model version vs new versions because parts of our distributed system will need to continue working with old version of data, embeddings and map image tile assets while the next version is being processed.
> 
> Hence we'll want something like a _build number_ or _model version_ to uniquely identify each along the way. [Similar to how OpenAI has tagged versions of GPT](https://platform.openai.com/docs/models/continuous-model-upgrades), etc.
> 
> I assumed that `_v01` was meant to track our embedding file format (which has already changed quite a few times if I'm not mistaken), or was that intended to track model version?

Gotcha, so we want a way to track both **model versions**, and **schema revisions** to the embedding file itself. I was maybe thinking of using `v01` to track both, but we could probably do either:

1. In the filename, have two parts like `{MODEL-VERSION}-{SCHEMA-REVISION}`, e.g. `_vit-1-abcde_v001.gpq`, where `vit-1-abcde` would mean Vision Transformer model 1, and `abcde` signifies the hash of the trained model, and `v001` being the schema version (e.g. what the columns and metadata inside the parquet file look like). The three digits could almost be SemVer-like, so 001 would mean schema v0.0.1.
2. Have just `{MODEL_VERSION}` in the filename e.g. `_vit-1-abcde.gpq`, and store the schema revision number internally in the Parquet file's metadata.

What do you think?

Reply:
Thanks for the quick back of the napkin math. While there probably isn't a hard ceiling on what constitutes too big for our primary use case (vector ingestion service), remember too that folks might want to open/work with these GeoParquet files from within jupyter notebooks or browser based javascript environments with resource limitations.

I would say yes let's include both `{MODEL-VERSION}`, `{SCHEMA-REVISION}` in parquet metadata for sure. It might also make sense to include `{MODEL-VERSION}` in the filename, or rather, the pathname like so:

`{outfolder}/{MODEL-VERSION}/{MGRS_CODE}.gpq`

We could add `DATE` to the pathname as well, perhaps after MODEL is best:

`{MODEL-VERSION}/{DATE}/{MGRS_CODE}.gpq`

I like your idea around date range. Hopefully it your code wouldn't have to do a table scan of all embeddings to determine MINDATE, MAXDATE... remember that could be 1.5G!

If date range is available / doable, that could look like:

`{MODEL-VERSION}/{MINDATE}_{MAXDATE}/{MGRS_CODE}.gpq`

Yeah, that's looking good to me!

Which begs the question - what are the typical use cases we foresee w.r.t. exporting vector embeddings? I imagine during development you might hard-code things to only work with a subset such as one-tile, but how do we expect to do things in production on a regular basis?

Would export be something done on demand by an engineer/operator from the command line or API request, etc. specifying parameters for desired MGRS tile(s) or date range?

Also, an aside, are we generating the GeoParquet file to scratch disk space and then uploading to S3? or is it generated in memory? The reason I ask is that when I make "file name" suggestions I am specifically thinking of the S3 object's file name to be clear. Feel free to name things on disk any which way you like or makes the most sense.

Reply:
Hmm... or does it make more sense for `{MGRS}` to come before `{DATE}` ? 

Not sure how often someone will be browsing the S3 bucket we use for ingestion to look for a specific MGRS tile's embedding but that might be a viable use case for data scientists in which case they know the MGRS code but not the date range a priori, in which case yeah, MGRS should probably come first left-to-right. 

`{MODEL-VERSION}/{MGRS}/{MINDATE}_{MAXDATE}.gpq`

Reply:
Sorry for the late reply. To summarize real quick, I think we should go for `{MGRS}_{MINDATE}_{MAXDATE}.gpq` for the filename would be best for now, and `{MODEL-VERSION}` will definitely be used at the folder level (i.e. a path like `{MODEL-VERSION}/{MGRS)_{MINDATE})_{MAXDATE}.gpq`. I'll need to have a think about whether to include the model version and schema revision in the filename too, but will do that in a separate PR.

> Which begs the question - what are the typical use cases we foresee w.r.t. exporting vector embeddings? I imagine during development you might hard-code things to only work with a subset such as one-tile, but how do we expect to do things in production on a regular basis?
> 
> Would export be something done on demand by an engineer/operator from the command line or API request, etc. specifying parameters for desired MGRS tile(s) or date range?
> 
> Also, an aside, are we generating the GeoParquet file to scratch disk space and then uploading to S3? or is it generated in memory? The reason I ask is that when I make "file name" suggestions I am specifically thinking of the S3 object's file name to be clear. Feel free to name things on disk any which way you like or makes the most sense.

Just to be clear, the neural network model is not likely something we update frequently, since it costs a lot of time and money to train a Foundation Model on lots of data. At such, you might only see a new model version every 3 months, or every 6 months, depending on what budget there is.

As for generating the embeddings (aka the embedding factory), this can be done in multiple ways, depending on what vector database you've decided upon, and what downstream applications you have in mind

- Currently we are processing each MGRS tile in batches, and doing a bulk export of embeddings to a GeoParquet file. For applications such as similarity search where you need a vector database of indexed embeddings to search over, a bulk batch method like this makes sense. E.g. if you want to find images of all the solar panels in the world.
- Alternatively, if you want to generate embeddings on demand via an API request, that may not even require setting up a vector databse capable of storing millions of rows. For example, if you want to visualize embeddings for 1 MGRS tile over 5 years in a time-series, it might be better to set up something like HuggingFace's [Inference API](https://huggingface.co/docs/api-inference/index) to do this, and just get a hundred rows out.

So my question is, are you looking at bulk scale embedding generation (100k+ rows), or small scale embedding generation (100s of rows)?